### DISCUSSION TITLE
Can't get @apply 'classname' for classname composition to work.

### DISCUSSION THREAD

**Discussion by herrKlein** [unocss/unocss#3555] [Q&A] [answered]
I would like to compose classes using apply with my own created classes. 
I cannot get this to work in UnoCSS, but this does work in tailwind:

```css
  .base { @apply m-4 }
  .button { @apply base } <-- use base class
```

I'm converting a tailwind project to unocss, which has more flexibility for our needs.
The project is a library, and we need to deliver an opinionated styled, but want people to copy these in their own project if they want to alter it. All our components are composed this way, but how to do this in unocss.. 

Playground here: 
[base composition example](https://uno.antfu.me/play/?html=DwEwlgbgBAxgNgQwM5ILwCIBGBXALrgewDt0A%2BAKCigCEEYBrAcwCcDsiQowkoiDcoAJQCmIcsAD04CKSA&config=PTAEFMGcBsEsDsAuBaAJrSBDARtcz5wAPFOQ0eAe2QCcpEbYBjRcVZWAWwAdKbFIAKC69%2BoBpniQAZn07gaAEVh0WsAG5RQ0mpU6gA5AAEArlSaRIwCVNk15NNCvBrNkA4OE8%2BiUAG9QVHBpBHAAYUp4EIBzABpQbjpIcEQAVSpQAF9tXX0DM0oLd09iUV8g6UwTaHLg0Iio2GiACj9BUHEaSRk5BUgALlAAbRse%2BwVlVUQNKGaASgBdWPaEpJSB4cSoFPTKeaXBTLnBIA&css=AIGwhgngpgTgBAIzAZynA3gKDnAdE1DbHOYMAB3JAjgGMBXGZAexgFpzmBLAOwBdYcGM3o8AJlDFsAtmLgi%2BIXlDY9mPNAgDmbGJLYBWAAxGA3MQC%2BmK5lCRBtZtM4b%2ByIjnz0%2BfdR5JklNSIKGjkAB5sAMxw5BBsAExwAGbqfGyo0lwIzCByyAAWYGLMAO4Z0uY4VhZAA&options=N4IgLgTghgdgzgMwPYQLYGECucxIwZXxAC4EoAbOAUwBpxp5k0AJMVcksyqgXyA)

```css
@layer base {
  .base {
    @apply cursor-pointer rounded-md border-2 border-blue-200 outline-none ring-1 ring-inset ring-gray-300;
  }
}

@layer components {
  .button {
    @apply base px-3 py-2 font-semibold shadow-sm;
  }
}
```

**Comment by herrKlein** [comment]
bump, still not possible, or am i missing something?

**Reply by zyyv** [reply, marked as answer]
UnoCSS does not support dynamically extracting CSS classes into atomic classes. You can try the `shortcut`.
[Playground](https://uno.antfu.me/play/#html=DwEwlgbgBAxgNgQwM5ILwCIBGBXALrgewDt0A%2BAKCigCEEYBrAcwCcDsiQowkoiDcoAJQCmIcsAD04CKSA&config=PTAEFMGcBsEsDsAuBaAJrSBDARtcz5wAPFOQ0eAe2QCcpEbYBjRcVZWAWwAdKbFIAKC69%2BoBpniQAZn07gaAEVh0WsAG5RQ0mpU6gA5AAEArlSaRIwCVNk15NNCvBrNkA4OE8%2BiUAG9QVHBpBHAAYUp4EIBzABpQbjpIcEQAVSpQAF9tXX0DM0oLd09iUV8g6UwTaHLg0Iio2GiACj9BUHEaSRk5BUgALlAAbRse%2BwVlVUQNKGaASgBdWPaEpJSB4cSoFPTKeaWVyAALHyYTAUG2jo7sTGTBgzOaSD5kXgRWGlBdMyD2TlQoEo5zI%2BCo5Gw0VobGQAFYAAzwjwdTKCTJzQRAA&css=AIGwhgngpgTgBAIzAZynA3gKDnAdE1DbHOYMAB3JAkRSgG5iBfTFzUSWOAYwHsBbcrwB2UYQBdkRHPgCu48SOkkylarULkAHgFoAzHHIQdAJjgAzEeJ2p%2BASwS8QAEzjIAFmGe8A7jf6MOCxMQA&options=N4IgLgTghgdgzgMwPYQLYGECucxIwZXxAC4EoAbOAUwBpxp5k0AJMVcksyqgXyA)
